libcudacxx/include/cuda/std/__execution/env.h

```diff
@@ -0,0 +1,319 @@
+ * collection of key/value pairs.
+ *
+ * This header defines templates and structures for querying properties of environments
+ * and managing them in a composable way. It includes support for querying properties of
+ * individual environments, combining multiple environments, and handling reference
+ * wrappers.
+ *
+ * An environment is a collection of key/value pairs where the key is an empty tag type.
+ * Environment objects have a `query` member function that accepts a key and returns the
+ * associated value.
+ *
+ * @details
+ * The key components in this file include:
+ * - `__query_result_t`: A type alias for the result of querying a query from an
+ *   environment.
+ * - `__queryable_with`: A concept to check if a query can be property from an
+ *   environment.
+ * - `__nothrow_queryable_with`: A concept to check if a property can be queried without
+ *   potentially throwing.
+ * - `prop`: Builds an environment from a key/value pair.
+ * - `env`: Builds an environment from a variadic number of environments. It allows
+ *   querying properties from the first environment that satisfies a given query type.
+ * - `get_env_t`: A callable object for retrieving the environment associated with an
+ *   object.
+ * - `env_of_t`: A type alias for the environment type of an object.
+ *
+ * @namespace cuda::std::execution
+ * The primary namespace for all components defined in this file.
+ *
+ * @concept __queryable_with
+ * Checks if a query `_Query` can be queried from an environment `_Ty`.
+ * @tparam _Ty The type of the environment.
+ * @tparam _Query The type of the property to be queried.
+ *
+ * @concept __nothrow_queryable_with Checks if a query `_Query` can be queried from an
+ * environment `_Ty` without potentially throwing.
+ * @tparam _Ty The type of the environment.
+ * @tparam _Query The type of the property to be queried.
+ *
+ * @struct prop
+ * @tparam _Query The type of the .
+ * @tparam _Value The type of the value associated with the query.
+ * A simple environment with a single value associated with a query.
+ *
+ * @struct env
+ * @tparam _Envs The types of the sub-environments contained within this environment.
+ * Represents a composable execution environment that can query properties from its
+ * sub-environments.
+ *
+ * @struct get_env_t
+ * A callable object for retrieving the environment associated with an object.
+ */
+
+#define _CCCL_API         _CCCL_HOST_DEVICE _CCCL_VISIBILITY_HIDDEN _CCCL_EXCLUDE_FROM_EXPLICIT_INSTANTIATION
+#define _CCCL_TRIVIAL_API _CCCL_API _CCCL_FORCEINLINE _CCCL_ARTIFICIAL _CCCL_NODEBUG
+
+_LIBCUDACXX_BEGIN_NAMESPACE_STD
+
+namespace detail
+{
+template <class _Ty, class _Query>
+auto __query_result_() -> decltype(declval<_Ty>().query(_Query()));
+
+#if _CCCL_HAS_EXCEPTIONS()
+template <class _Ty, class _Query>
+using __nothrow_queryable_with_ _CCCL_NODEBUG_ALIAS = enable_if_t<noexcept(declval<_Ty>().query(_Query{}))>;
+#endif
```

Comment:
With exceptions of shouldnt this rather be set to unconditionally be true?

Reply:
no why?

Reply:
Sorry, *without* exceptions we could just set this as true, but I see you did that below